spec/abci++/abci++_methods_002_draft.md

```diff
@@ -0,0 +1,839 @@
+        * reorder transactions - the Application reorders transactions in the list
+4. If the block is modified, the Application sets `ResponsePrepareProposal.modified` to true,
+   and includes the modified block in the return parameters (see the rules in section _Usage_).
+   The Application returns from the call.
+5. _p_'s Tendermint uses the (possibly) modified block as _p_'s proposal in round _r_, height _h_.
+
+Note that, if _p_ has a non-`nil` _validValue_, Tendermint will use it as proposal and will not call `RequestPrepareProposal`.
+
+### ProcessProposal
+
+#### Parameters and Types
+
+* **Request**:
+
+    | Name                 | Type                                        | Description                                                                                                      | Field Number |
+    |----------------------|---------------------------------------------|------------------------------------------------------------------------------------------------------------------|--------------|
+    | hash                 | bytes                                       | The block header's hash of the proposed block. Present for convenience (can be derived from the block header).   | 1            |
+    | header               | [Header](../core/data_structures.md#header) | The proposed block's header.                                                                                     | 2            |
+    | tx                   | repeated bytes                              | List of transactions that have been picked as part of the proposed block.                                        | 3            |
+    | byzantine_validators | repeated [Evidence](#evidence)              | List of evidence of validators that acted maliciously.                                                           | 4            |
+    | last_commit_info     | [LastCommitInfo](#lastcommitinfo)           | Info about the last commit, including the round , the validator list, and which ones signed the last block.      | 5            |
+
+* **Response**:
+
+    | Name   | Type | Description                                      | Field Number |
+    |--------|------|--------------------------------------------------|--------------|
+    | accept | bool | If false, the received block failed verification | 1            |
+
+* **Usage**:
+    * Contains a full proposed block.
+        * The parameters and types of `RequestProcessProposal` are the same as `RequestPrepareProposal`
+          and `RequestFinalizeBlock`.
+        * The Application may fully execute the block as though it was handling `RequestFinalizeBlock`.
+          However, any resulting state changes must be kept as _canditade state_,
+          and the Application should be ready to backtrack/discard it in case the decided block is different.
+    * The header contains the height, timestamp, and more - it exactly matches the
```

Comment:
It's not fully clear for me what will be in the `Header.AppHash` (and similar fields). Will it be `ResponsePrepareProposal.data` for the same height? If not, how can the non-proposer validate `ResponsePrepareProposal.data`?